### ISSUE TITLE
Extraction framework

### ISSUE THREAD

**Issue by kstaken** [terascope/alerting-assets#1]
In addition to simple selector based alerts we want to be able to build alerts and other workflows that extract pieces from the incoming documents and build new result documents.

In a normal alerting workflow you would have 2 phases.

1. Selection - Applies a list of xlucene selectors to the incoming data flow
2. Output - Feeds the next phase of the pipeline with the selected documents.

In the extraction workflow there are 4 phases.

1. Selection - this is basically the same for any alert list.
2. Extraction - Applies extraction rules to selected documents and builds up a new record.
3. Validation - Takes the generated document and runs it against defined validation rules.
4. Output - Feeds the next phase of the pipeline with the the generated document.

We should not assume anything about how the output of extraction will be used. The documents could be fed into the standard alert `actions` logic or they may just be the first step of a generic Teraslice pipeline.

### Extraction rules.

Extraction rules might look something like the following.
```
{
     selector: "field1:value AND field2:value",
     source_field: "myfield",
     start: "field1=",
     end: "EOP",
     target_field: "topfield.value1"
}
```

The rules would be specified in a list, one record per line. NOTE: That obviously implies they will not have the indention as shown above.

#### Field definitions
 * `selector` - The selector is an `xlucene` selector that will be applied to the incoming stream to filter documents for extraction.  If no `selector` is provided then all records will pas through the extraction logic. `OPTIONAL`
 * `source_field` - This is the field in the `incoming` document` that contains the source string to extract data from. `REQUIRED`
 * `target_field` - The extracted value if any will be placed in this field in the generated document. NICE TO HAVE: Allow path expressions to create more complex documents. `field` or `field.subfield` or `field.subfield[].field`.  `REQUIRED`
 * `start` - The pattern that defines the start of a field. This field is `OPTIONAL`
 * `end`  - The pattern that defines the end of a field or a special of field marker. `EOP` means end of property and for URL extractions would match `&` or the end of the line. There will likely need to be a small selection of built in patterns for this but the end user is free to specify a generic string. `OPTIONAL`
 * `regex` Allows the user to provide a full regex pattern in lieu of `start` and `end`. `OPTIONAL`
 * `validation` Validation logic defined on a field but applied to the end document. One particular usage of this will be to extract multiple values from independent rules and then combine them together as a single entity. See example below.`OPTIONAL`
 * `post_process` - An optional post_processing function for the extracted value. This would be something like `base64decode`. We should expect a small library of these to be developed. This is a shorthand helper for the more complicated operations chains described below in the case where this is a single simple transform. `OPTIONAL` 

#### Example with a validation rule

This example contains two rules that extract two fields from the input to create  a more complicated object in the output.

```
{
     selector: "hostname:www.example.com",
     source_field: "path",
     regex: "/path/tiles/latitude/(.*?)$"
     target_field: "location.lat",
     validation: "geolocation"
}
{
     selector: "hostname:www.example.com",
     source_field: "path",
     regex: "/path/tiles/longitude/(.*?)$"
     target_field: "location.lon",
     validation: "geolocation"
}
```

Those two rules would combine to create a document like this:

```
{
    location: {
         lon: 29.34,
         lat: -195.65
   }
}
```

Both rules define a geolocation validation that applies to the `location` field.  That validation would then verify that the extracted values are valid geo coordinates. If they are not valid the document would be rejected from being passed to the output phase.

We should expect a small library of validation rules will be required and there may be some existing libs we can build on in this area.

**Comment by kstaken** [comment]
We need to extend this to accommodate a couple more complicated scenarios.

I think this will introduce a fifth phase.

1. selection
2. extraction
3. post-processing
4. validation
5. output

The post processing phase will take the initial output of extraction and then perform additional operations on the data. The operations that are defined should be an open framework.

Operations in the validation phase will actually be pretty similar it's just a difference on when they are applied.

Post processing operations can be specified two ways.

1. Attached to an existing rule via an `id`
```
{
     selector: "hostname:www.example.com",
     source_field: "path",
     regex: "/path/tiles/latitude/(.*?)$",
     target_field: "location.lat",
     validation: "geolocation",
     id: "location-lat"
}
{
     refs: "location-lat",
     operation: "base64decode",
     target_field: "location.lat"
}
```
  * In this scenario the `base64decode` would apply to the `target_field` from the chained operation and then overwrite that same field
  *  The set of parameters for each operation will be defined by the specific operation.
  *  If there are multiple operations defined referencing the same ID then they should be applied in the order they are defined.
  * `id`s have to be unique across all rules in a given rule set.
  * The operations are a chain. If an `id` is also defined on the `base64decode` operation then it should be possible to attach additional operations to follow that operation.
  * The complete chain of operations should be completed before moving to the next step. i.e. you should complete the entire chain horizontally before moving to the next step vertically and the input to that next step would be the output of the entire horizontal chain.
```
operation1 -> operation2 -> operation3
       |
operation4 (get's doc as output from operation3)
```

NOTE: The example above is a more generalized form of the following.

```
{
     selector: "hostname:www.example.com",
     source_field: "path",
     regex: "/path/tiles/latitude/(.*?)$",
     target_field: "location.lat",
     validation: "geolocation",
     post_process: "base64decode"
}
```

Here post_process was what I originally called `decoder`. This is purely syntax sugar to make simple cases easier to write.

It is absolutely critical that it is clear that these two syntaxes translate into the EXACT SAME execution mechanism. If this is done the framework will be powerful and extensible. 

2. Attached to a selector
```
{
     selector: "hostname:www.example.com",
     operation: "join",
     fields: ["first_name", "last_name"],
     delimiter: " ",
     target_field: "full_name",
     remove_source: true
}
```

* In this instance the operation is added into the post_process pipeline directly and not attached to a particular extraction.
* It should be added at the end of the pipeline of post_processors and would apply to the entire generated document so that mutations could be applied to that document.
* In multiple post_processors are defined this way then they should be applied in the order they are defined.

Obviously we will also build a library of these operations for transforming the data. In reality all extractions/transformations are just applying this same model to different inputs. So an extraction could be written as.

```
{
     selector: "hostname:www.example.com",
     operation: "extraction",
     source_field: "path",
     regex: "/path/tiles/latitude/(.*?)$"
     target_field: "location.lat",
     validation: "geolocation"
}
```

We just implicitly assume that if `operation` is not specified then it defaults to `extraction` or whatever it is called.

**Comment by kstaken** [comment]
As an additional refinement we decided to rename `operation` in post process rules to `post_process`. This makes it consistent with the already defined `post_process` attribute on extractions.

```
{
     selector: "hostname:www.example.com",
     source_field: "path",
     regex: "/path/tiles/latitude/(.*?)$",
     target_field: "location.lat",
     validation: "geolocation",
     id: "location-lat"
}
{
     refs: "location-lat",
     post_process: "base64decode",
     target_field: "location.lat"
}
```

And this are two ways of expressing the same thing.

```
{
     selector: "hostname:www.example.com",
     source_field: "path",
     regex: "/path/tiles/latitude/(.*?)$",
     target_field: "location.lat",
     validation: "geolocation",
     post_process: "base64decode"
}
```

We also discussed making sure that rules can be defined without selectors. If a rule has no selector then it would be considered basically `selector='*'` and all records would pass through the associated extractions.

```
{
     source_field: "path",
     regex: "/path/tiles/(.*?)$",
     target_field: "filename"
}
```

**Comment by kstaken** [comment]
One other thought, we discussed the idea of implementing the separate phases as different processors. As discussed it was just a concept to explain the model but it's actually something we should seriously consider as an option. We wouldn't want it to be the only way to use the framework but it opens the door to being able to use this framework in other contexts. 

For instance you could put additional processors between phases or use just use the validation phase or have multiple selection phases. 

```
selection 
     |
extraction
     |
selection <- runs on the extracted documents to further restrict the output based on different rules
     |
post_process <- could use the original rules or new rules
     |
validation
     |
output

```
Definitely don't want it to be the only way, but if architected this way the underlying code can be exposed through different processor interfaces for different use cases.

**Comment by kstaken** [comment]
Additional operations (this list will evolve)

#### Post process
* base64decode
* urldecode
* hexdecode
* join - combine fields

#### Validation - may be existing libs that will help here
* number 
* string
* boolean
* email
* url
* geolocation
* ip address (IPv4 / IPv6)

In addition to these it's pretty clear we're going to need to allow the usage of custom post_process/validation operations. I'd assume they would just be node modules that may define multiple operations as a library loaded by the processor.

**Comment by kstaken** [comment]
Just as we defined post_process above, we should define validation the same way.

```
{
     selector: "hostname:www.example.com",
     source_field: "path",
     regex: "/path/tiles/latitude/(.*?)$",
     target_field: "location.lat",
     id: "location-lat"
}
{
     refs: "location-lat",
     validation: "geolocation",
     target_field: "location"
}
```

As exactly the same thing as ...
```
{
     selector: "hostname:www.example.com",
     source_field: "path",
     regex: "/path/tiles/latitude/(.*?)$",
     target_field: "location.lat",
     validation: "geolocation"
}
```

This opens the door to validations that take additional parameters. In this example the string validation will require a filename that is 10 characters long.

```
{
     selector: "hostname:www.example.com",
     source_field: "path",
     regex: "/path/tiles/(.*?)$",
     target_field: "filename",
     id: "filename-rule"
}
{
     refs: "filename-rule",
     validation: "string",
     length: 10
}
```

This also exhibits that on post_process / validation if the target_field isn't specified it should default to the field providing the value being processed.

**Comment by jsnoble** [comment]
This is referencing ts-transforms
